pkg/driver/controller.go

```diff
@@ -602,7 +620,23 @@ func (d *controllerService) ControllerExpandVolume(ctx context.Context, req *csi
 
 func (d *controllerService) ControllerModifyVolume(ctx context.Context, req *csi.ControllerModifyVolumeRequest) (*csi.ControllerModifyVolumeResponse, error) {
 	klog.V(4).InfoS("ControllerModifyVolume: called", "args", *req)
-	return nil, status.Error(codes.Unimplemented, "")
+
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID not provided")
+	}
+
+	options, err := parseModifyVolumeParameters(req.GetMutableParameters())
```

Comment:
> Specifically, say a user provides an invalid mutable configuration such as gp2 and throughput - what would be the best strategy to handle that scenario: error out before making the backend call, or return the error from the server? I believe CSI spec is steering us towards the former.

I disagree with the former solution. 

The more driver-side validation we attempt, the more tightly we couple ourselves to a single version of the EC2 API, and the more likely we are to introduce a bug if the EC2 API changes (ie gp2 magically now has a throughput parameter)

The CSI Spec allows us to now declare a volume modification as invalid via the new error code. We should propagate the EC2 client error to the sidecar via this error code. But I think more driver-side validation is dangerous, as we learned with the GP3 min iops regression.

Reply:
Link to the infeasible status VAC-ModifyVolume flow: [enhancements/keps/sig-storage/3751-volume-attributes-class/VolumeAttributesClass-ModifyVolume-Flow-v3.png at master · kubernetes/enhancements](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/VolumeAttributesClass-ModifyVolume-Flow-v3.png)

Link to aforementioned regression: [Remove regression for GP3 storage classes below 3000 IOPS #1879](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1879)

Reply:
That is a typo in the CSI spec, the `ControllerModifyVolume` call does not receive capabilities and mutable parameters cannot be verified by the caller except by attempting a CMV call. There are no cases in which our CMV implementation returns an INVALID_ARGUMENT error for capability reasons.

The actual line for this error in the spec is
```
Parameters not supported 	3 INVALID_ARGUMENT 	Indicates that the CO has specified mutable parameters not supported by the volume. 	Caller MAY verify mutable parameters.
```

Reply:
I largely agree with that model -- if I could wave a magic wand and reduce the cost of backend calls to 0 then fully agree.

> an extra backend call is the price we have to pay for not being tightly coupled to today's EC2 API.

It's not just one extra backend call, it's n number of extra calls because the call will be retried using an exponential backoff strategy. All of those calls are technically unnecessary because (today) we know they will fail before the call is made.

If we acknowledged that CSI spec explicit recommendation of:

> Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities

is not being followed because we are willing to incur the cost of making the calls to account for the scenario where EC2 API changes and undesirable coupling as you point out, sgtm.

Reply:
Resolving as agreement has been reached - we will continue with our normal policy of not attempting to validate modify parameters beyond what is strictly necessary (confirming the string provided decodes to an integer).

Reply:
> There are no cases in which our CMV implementation returns an INVALID_ARGUMENT error for capability reasons.

Un-resolving this because we **must** return the `INVALID_ARGUMENT` gRPC code in case of an `Client.InvalidParameterValue` EC2 ModifyVolume call. 

Not doing so leads to many more wasteful EC2 API calls than if we return  `Client.InvalidParameterValue` due to the sidecar exponentially backing off instead of re-queueing after a much longer backoff as mentioned in the linked resource above: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3751-volume-attributes-class/VolumeAttributesClass-ModifyVolume-Flow-v3.png. 

What are the capability reasons that stop us from this? All we have to do is translate an AWS `Client.InvalidParameterValue` into the CSI `INVALID_ARGUMENT` error code from within ControllerModifyVolume?

Reply:
Discussed offline, this will be a separate PR. 

Must convert AWS Client.InvalidParameterValue -> Cloud package InvalidParameter error -> CSI INVALID_ARGUMENT